server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -531,6 +537,58 @@ public String readString() throws IOException {
         return new String(charBuffer, 0, charCount);
     }
 
+    protected String tryReadStringFromBytes(final byte[] bytes, final int start, final int limit, final int chars) throws IOException {
+        final int end = start + chars;
+        if (limit < end) {
+            return null; // not enough bytes to read chars
+        }
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0x80) != 0) {
+                // not an ASCII char, fall back to reading a UTF-8 string
+                return tryReadUtf8StringFromBytes(bytes, start, limit, pos, end - pos);
+            }
+        }
+        skip(chars); // skip the number of chars (equals bytes) on the stream input
+        return new String(bytes, start, chars, ISO_8859_1);
```

Comment:
shouldn't this encoding be ASCII? ISO-8859-1 is not a subset of the encoding for utf8 (the top bit may be set). Effectively it won't matter since ASCII is a subset of ISO-8859-1, but we should set the right encoding.

Reply:
@DaveCTurner brought this up, using `ISO-8859-1` safes some additional checks compared to `UTF_8`, but also a few compared to `US_ASCII`. By this line we've already check each byte is within [0, 127], so `ISO-8859-1` is safe to use and internally the coder is `LATIN1` in either case.

Reply:
Oh, I missed your comment @DaveCTurner . I'll add a comment on that line 👍